### PR TITLE
support separator string for iteration string param

### LINKED ISSUE #788: Pipelineloop should support to iterate with an string with separator char
/kind feature

**Description:**
Current pipeline loop only supports to iterate from `iterateParam` with string value like:
```
value: '["I am the first one", "I am the second one", "I am the third one"]'
```

I would like to introduce a new field `iterateParamSeparator` to allow user to pass the separator char and allow us to split the iterateParam as a string array for the loop.

This field is optional, it only works with string type param.

eg: 
```
iterateParam= "aaa|bbb|ccc|ddd"
iterateParamSeparator="|"
```

### PR DESCRIPTION
**Pull request by wzhanw**
**Which issue is resolved by this Pull Request:** 
Resolves #788 

**Description of your changes:**
Add a new optional field `IterateParamStringSeparator` to allow the user to specify the parameter for the field.

in reconcile logic, if the `iterationParam` is a string, we will check the `IterateParamStringSeparator` meanwhile, if it was configured, then try to get the value from the parameter.  Then split the string for iteration.

**Environment tested:**
* Tekton Version (use `tkn version`):
```
Client version: 0.17.2
Pipeline version: v0.27.3
Triggers version: v0.16.0
```
* Kubernetes Version (use `kubectl version`): 
``` Server Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.1", GitCommit:"206bcadf021e76c27513500ca24182692aabd17e", GitTreeState:"clean", BuildDate:"2020-09-14T07:30:52Z", GoVersion:"go1.15", Compiler:"gc", Platform:"linux/amd64"}
````
* OS (e.g. from `/etc/os-release`):

**Checklist:**
- [X] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### CODE REVIEW

File: tekton-catalog/pipeline-loops/pkg/apis/pipelineloop/v1alpha1/pipelineloop_types.go [hunk 1]
```diff
@@ -51,6 +51,10 @@ type PipelineLoopSpec struct {
 	// IterateParam is the name of the task parameter that is iterated upon.
 	IterateParam string `json:"iterateParam"`
 
+	// The separator for IterateParam if the IterateParam is a strings with separator char, this field is optional.
+	// +optional
+	IterateParamStringSeparator string `json:"iterateParamStringSeparator,omitempty"`
```
Comment (pugangxa) [inline tekton-catalog/pipeline-loops/pkg/apis/pipelineloop/v1alpha1/pipelineloop_types.go:56]: How about call it IterateParamSeparator?

File: tekton-catalog/pipeline-loops/pkg/reconciler/pipelinelooprun/pipelinelooprun.go [hunk 2]
```diff
@@ -597,53 +601,64 @@ func computeIterations(run *v1alpha1.Run, tls *pipelineloopv1alpha1.PipelineLoop
...
-				for _, v := range p.Value.ArrayVal {
+			if p.Name == tls.IterateParamStringSeparator {
+				iterationParamStrSeparator = p.Value.StringVal
+			}
+		}
+	}
+	if iterationParamStr != "" {
+		// Transfer p.Value to Array.
+		if iterationParamStrSeparator != "" {
+			stringArr := strings.Split(iterationParamStr, iterationParamStrSeparator)
```
Comment (pugangxa) [inline tekton-catalog/pipeline-loops/pkg/reconciler/pipelinelooprun/pipelinelooprun.go:625]: Should we strip the string?

File: tekton-catalog/pipeline-loops/pkg/reconciler/pipelinelooprun/pipelinelooprun.go [hunk 3]
```diff
@@ -597,53 +601,64 @@ func computeIterations(run *v1alpha1.Run, tls *pipelineloopv1alpha1.PipelineLoop
...
 			}
-			if p.Value.Type == v1beta1.ParamTypeArray {
-				numberOfIterations = len(p.Value.ArrayVal)
-				for _, v := range p.Value.ArrayVal {
+			if p.Name == tls.IterateParamStringSeparator {
+				iterationParamStrSeparator = p.Value.StringVal
+			}
+		}
+	}
+	if iterationParamStr != "" {
```
Comment (pugangxa) [inline tekton-catalog/pipeline-loops/pkg/reconciler/pipelinelooprun/pipelinelooprun.go:622]: Is white space allowed? In yaml it should be " "?

**Comment by pugangxa** [comment]
Question: How to use it? It's just the backend changes, we need a DSL interface so that user can define it in python.

**Comment by Udiknedormin** [comment]
@pugangxa Is there any support for numeric range in DSL? It seems to me that as long as there is no DSL extension for PipelineLoop in general, only the KFP's dsl.ParallelFor, then it's beyond the scope of this PR to add it. It can be handled separately, I believe.

**Comment by pugangxa** [comment]
> @pugangxa Is there any support for numeric range in DSL? It seems to me that as long as there is no DSL extension for PipelineLoop in general, only the KFP's dsl.ParallelFor, then it's beyond the scope of this PR to add it. It can be handled separately, I believe.

Yes there's no DSL extension for PipelineLoop, just think without it user who want to try this feature they must edit yaml. It's OK to be handled later though.

File: tekton-catalog/pipeline-loops/pkg/reconciler/pipelinelooprun/pipelinelooprun.go (hunk 3 above)
Comment (wzhanw) [reply tekton-catalog/pipeline-loops/pkg/reconciler/pipelinelooprun/pipelinelooprun.go:622]: allowed, I added the unit test case

File: tekton-catalog/pipeline-loops/pkg/reconciler/pipelinelooprun/pipelinelooprun.go (hunk 2 above)
Comment (wzhanw) [reply tekton-catalog/pipeline-loops/pkg/reconciler/pipelinelooprun/pipelinelooprun.go:625]: add the code to trim the string first.

File: tekton-catalog/pipeline-loops/pkg/apis/pipelineloop/v1alpha1/pipelineloop_types.go (hunk 1 above)
Comment (wzhanw) [reply tekton-catalog/pipeline-loops/pkg/apis/pipelineloop/v1alpha1/pipelineloop_types.go]: changed to `IterateParamSeparator `

**Pull request merged**